File: 199_Binary_Tree_Right_Side_View.cpp
/*
Given a binary tree, imagine yourself standing on the right side of it, 
return the values of the nodes you can see ordered from top to bottom.

Example:

Input: [1,2,3,null,5,null,4]
Output: [1, 3, 4]
Explanation:

   1            <---
 /   \
2     3         <---
 \     \
  5     4       <---
*/
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
    public:
        struct TreeNode {
            int val;
            TreeNode *left;
            TreeNode *right;
            TreeNode(int x) : val(x), left(NULL), right(NULL) {}
        };

        vector<int> rightSideView(TreeNode* root) {
            vector<int> res;
            BFS(root, 1, res);
            return res;
        }

        void BFS(TreeNode* root, int depth, vector<int>& result) {
            if(!root) return;
            if(result.size() < depth) result.push_back(root->val);
            BFS(root->right, depth + 1, result);
            BFS(root->left, depth + 1, result);
        }
};

int main() {
    return 0;
}

/*


*/

File: py/199_Binary_Tree_Right_Side_View.py
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        res = []
        self.bfs(root, 1, res)
        return res

    def bfs(self, root: TreeNode, depth: int, result: List[int]):
        if not root:
            return
        if len(result) < depth:
            result.append(root.val)
        self.bfs(root.right, depth + 1, result)
        self.bfs(root.left, depth + 1, result)
